### PR TITLE
Allow moby-buildx version to be specified in devcontainer.json

### LINKED ISSUE #837: Ubuntu:focal + latest docker-in-docker feature fails to prebuild due to Moby package validation issue
Hi there,

I manage a devcontainer config for Codespaces that uses the `ubuntu:focal` image (like the universal image does) alongside the latest docker-in-docker feature. As of around 10am Pacific time today, our Codespaces prebuilds began to fail with the following error:

```shell
#28 18.36 Err:1 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-buildx amd64 0.12.1-ubuntu20.04u1
#28 18.36   File has unexpected size (34152700 != 34091250). Mirror sync in progress? [IP: 13.90.21.104 443]
```

GitHub Actions Prebuild logs:
```shell
#28 0.707 ===========================================================================
#28 0.707 Feature       : Docker (Docker-in-Docker)
#28 0.707 Description   : Create child containers *inside* a container, independent from the host's docker instance. Installs Docker extension in the container along with needed CLIs.
#28 0.707 Id            : ghcr.io/devcontainers/features/docker-in-docker
#28 0.707 Version       : 2.8.1
#28 0.707 Documentation : ********/devcontainers/features/tree/main/src/docker-in-docker
#28 0.707 Options       :
#28 0.707     VERSION="latest"
#28 0.707     MOBY="********"
#28 0.707     DOCKERDASHCOMPOSEVERSION="v1"
#28 0.707     AZUREDNSAUTODETECTION="********"
#28 0.707     DOCKERDEFAULTADDRESSPOOL=""
#28 0.707     INSTALLDOCKERBUILDX="********"
#28 18.36 Err:1 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-buildx amd64 0.12.1-ubuntu20.04u1
#28 18.36   File has unexpected size (34152700 != 34091250). Mirror sync in progress? [IP: 13.90.21.104 443]
#28 18.36   Hashes of expected file:
#28 18.36    - SHA256:0d1c7ac216d2a7d825b6fb0dbdec0f81d88540f1cc074d77abf4c0e7988f3e03
#28 18.36    - Filesize:34091250 [weak]
#28 18.38 Get:2 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-cli amd64 24.0.9-ubuntu20.04u1 [15.6 MB]#28 18.58 Get:3 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-runc amd64 1.1.12-ubuntu20.04u1 [6680 kB]#28 18.64 Get:4 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-containerd amd64 1.6.28-ubuntu20.04u1 [45.9 MB]
#28 18.99 Get:5 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-tini amd64 0.19.0-ubuntu20.04u1 [350 kB]
#28 18.99 Get:6 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-engine amd64 24.0.9-ubuntu20.04u1 [32.2 MB]
#28 19.22 Fetched 101 MB in 1s (113 MB/s)
#28 19.22 E: Failed to fetch https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod/pool/main/m/moby-buildx/moby-buildx_0.12.1-ubuntu20.04u1_amd64.deb  File has unexpected size (34152700 != 34091250). Mirror sync in progress? [IP: 13.90.21.104 443]
#28 19.22    Hashes of expected file:
#28 19.22     - SHA256:0d1c7ac216d2a7d825b6fb0dbdec0f81d88540f1cc074d77abf4c0e7988f3e03
#28 19.22     - Filesize:34091250 [weak]
#28 19.22 E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
#28 19.23 (!) Packages for moby not available in OS ubuntu focal (amd64). To resolve, either: (1) set feature option '"moby": false' , or (2) choose a compatible OS version (eg: 'ubuntu-20.04').
#28 19.24 ERROR: Feature "Docker (Docker-in-Docker)" (ghcr.io/devcontainers/features/docker-in-docker) failed to install! Look at the documentation at ********/devcontainers/features/tree/main/src/docker-in-docker for help troubleshooting this error.
#28 ERROR: process "/bin/sh -c cp -ar /tmp/build-features-src/docker-in-docker_6 /tmp/dev-container-features  && chmod -R 0755 /tmp/dev-container-features/docker-in-docker_6  && cd /tmp/dev-container-features/docker-in-docker_6  && chmod +x ./devcontainer-features-install.sh  && ./devcontainer-features-install.sh  && rm -rf /tmp/dev-container-features/docker-in-docker_6" did not complete successfully: exit code: 1
------
 > [dev_containers_target_stage 10/10] RUN --mount=type=bind,from=dev_containers_feature_content_source,source=docker-in-docker_6,target=/tmp/build-features-src/docker-in-docker_6     cp -ar /tmp/build-features-src/docker-in-docker_6 /tmp/dev-container-features  && chmod -R 0755 /tmp/dev-container-features/docker-in-docker_6  && cd /tmp/dev-container-features/docker-in-docker_6  && chmod +x ./devcontainer-features-install.sh  && ./devcontainer-features-install.sh  && rm -rf /tmp/dev-container-features/docker-in-docker_6:
18.99 Get:5 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-tini amd64 0.19.0-ubuntu20.04u1 [350 kB]
18.99 Get:6 https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod focal/main amd64 moby-engine amd64 24.0.9-ubuntu20.04u1 [32.2 MB]
19.22 Fetched 101 MB in 1s (113 MB/s)
19.22 E: Failed to fetch https://packages.microsoft.com/repos/microsoft-ubuntu-focal-prod/pool/main/m/moby-buildx/moby-buildx_0.12.1-ubuntu20.04u1_amd64.deb  File has unexpected size (34152700 != 34091250). Mirror sync in progress? [IP: 13.90.21.104 443]
19.22    Hashes of expected file:
19.22     - SHA256:0d1c7ac216d2a7d825b6fb0dbdec0f81d88540f1cc074d77abf4c0e7988f3e03
19.22     - Filesize:34091250 [weak]
19.22 E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
19.23 (!) Packages for moby not available in OS ubuntu focal (amd64). To resolve, either: (1) set feature option '"moby": false' , or (2) choose a compatible OS version (eg: 'ubuntu-20.04').
19.24 ERROR: Feature "Docker (Docker-in-Docker)" (ghcr.io/devcontainers/features/docker-in-docker) failed to install! Look at the documentation at ********/devcontainers/features/tree/main/src/docker-in-docker for help troubleshooting this error.
------
Dockerfile-with-features:183
--------------------
 182 |     ENV DOCKER_BUILDKIT="1"
 183 | >>> RUN --mount=type=bind,from=dev_containers_feature_content_source,source=docker-in-docker_6,target=/tmp/build-features-src/docker-in-docker_6 \
 184 | >>>     cp -ar /tmp/build-features-src/docker-in-docker_6 /tmp/dev-container-features \
 185 | >>>  && chmod -R 0755 /tmp/dev-container-features/docker-in-docker_6 \
 186 | >>>  && cd /tmp/dev-container-features/docker-in-docker_6 \
 187 | >>>  && chmod +x ./devcontainer-features-install.sh \
 188 | >>>  && ./devcontainer-features-install.sh \
 189 | >>>  && rm -rf /tmp/dev-container-features/docker-in-docker_6
 190 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c cp -ar /tmp/build-features-src/docker-in-docker_6 /tmp/dev-container-features  && chmod -R 0755 /tmp/dev-container-features/docker-in-docker_6  && cd /tmp/dev-container-features/docker-in-docker_6  && chmod +x ./devcontainer-features-install.sh  && ./devcontainer-features-install.sh  && rm -rf /tmp/dev-container-features/docker-in-docker_6" did not complete successfully: exit code: 1
Stop: Run: docker buildx build --load --build-arg BUILDKIT_INLINE_CACHE=1 -f /tmp/devcontainercli-root/container-features/0.56.1-1707501892313/Dockerfile-with-features -t vsc-web-abc123 --target dev_containers_target_stage --build-context dev_containers_feature_content_source=/tmp/devcontainercli-root/container-features/0.56.1-1707501892313 --build-arg _DEV_CONTAINERS_BASE_IMAGE=dev_container_auto_added_stage_label --build-arg _DEV_CONTAINERS_IMAGE_USER=root --build-arg _DEV_CONTAINERS_FEATURE_CONTENT_SOURCE=dev_container_feature_content_temp /var/lib/docker/codespacemount/workspace/web/.devcontainer
Error: Command failed: docker buildx build --load --build-arg BUILDKIT_INLINE_CACHE=1 -f /tmp/devcontainercli-root/container-features/0.56.1-1707501892313/Dockerfile-with-features -t vsc-web-abc123 --target dev_containers_target_stage --build-context dev_containers_feature_content_source=/tmp/devcontainercli-root/container-features/0.56.1-1707501892313 --build-arg _DEV_CONTAINERS_BASE_IMAGE=dev_container_auto_added_stage_label --build-arg _DEV_CONTAINERS_IMAGE_USER=root --build-arg _DEV_CONTAINERS_FEATURE_CONTENT_SOURCE=dev_container_feature_content_temp /var/lib/docker/codespacemount/workspace/web/.devcontainer
    at j$ (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:465:1933)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Lw (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:464:1831)
    at async iK (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:464:608)
    at async gAA (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:481:3660)
{"outcome":"error","message":"Command failed: docker buildx build --load --build-arg BUILDKIT_INLINE_CACHE=1 -f /tmp/devcontainercli-root/container-features/0.56.1-1707501892313/Dockerfile-with-features -t vsc-web-abc123 --target dev_containers_target_stage --build-context dev_containers_feature_content_source=/tmp/devcontainercli-root/container-features/0.56.1-1707501892313 --build-arg _DEV_CONTAINERS_BASE_IMAGE=dev_container_auto_added_stage_label --build-arg _DEV_CONTAINERS_IMAGE_USER=root --build-arg _DEV_CONTAINERS_FEATURE_CONTENT_SOURCE=dev_container_feature_content_temp /var/lib/docker/codespacemount/workspace/web/.devcontainer","description":"An error occurred building the image."}
    at async BC (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:481:4775)
    at async xeA (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:614:11265)
    at async UeA (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:614:11006)
devcontainer process exited with exit code 1
Failed to create container.
Error: Command failed: docker buildx build --load --build-arg BUILDKIT_INLINE_CACHE=1 -f /tmp/devcontainercli-root/container-features/0.56.1-1707501892313/Dockerfile-with-features -t vsc-web-abc123 --target dev_containers_target_stage --build-context dev_containers_feature_content_source=/tmp/devcontainercli-root/container-features/0.56.1-1707501892313 --build-arg _DEV_CONTAINERS_BASE_IMAGE=dev_container_auto_added_stage_label --build-arg _DEV_CONTAINERS_IMAGE_USER=root --build-arg _DEV_CONTAINERS_FEATURE_CONTENT_SOURCE=dev_container_feature_content_temp /var/lib/docker/codespacemount/workspace/web/.devcontainer
Error code: 1302 (UnifiedContainersErrorFatalCreatingContainer)
Container creation failed.
Jobs failed, exiting the agent. Error code: 1302 (UnifiedContainersErrorFatalCreatingContainer)
Error: Process completed with exit code 1.
```

devcontainer.json:
```json
 "build": {
    "dockerfile": "./Dockerfile",
    "context": "."
  },
  "runArgs": [ "--init", "--privileged" ], // for Docker-in-Docker
  "overrideCommand": false, // for Docker-in-Docker

  "features": {
    "ghcr.io/devcontainers/features/common-utils:2": {
      "username": "codespace",
      "userUid": "1000",
      "userGid": "1000"
    },
    "ghcr.io/devcontainers/features/docker-in-docker:2": {
      "version": "latest"
    },
    ...
  },

  "overrideFeatureInstallOrder": [
    "ghcr.io/devcontainers/features/common-utils",
    ...
    "ghcr.io/devcontainers/features/docker-in-docker",
  ],
```

Dockerfile:
```dockerfile
FROM ubuntu:focal

COPY first-run-notice.txt /tmp/scripts/

RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
    # Restore man command
    && yes | unminimize 2>&1

ENV LANG="C.UTF-8"

# Install basic build tools
RUN apt-get update \
    && apt-get upgrade -y \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
    [all the stuff from the universal image]

# Verify expected build and debug tools are present
RUN apt-get update \
    && apt-get -y install build-essential cmake cppcheck valgrind clang lldb llvm gdb python3-dev \
    # Install tools and shells not in common script
    && apt-get install -yq vim vim-doc xtail software-properties-common libsecret-1-dev \
    # Install additional tools (useful for 'puppeteer' project)
    && apt-get install -y --no-install-recommends l
    [stuff from universal]

# Default to bash shell (zsh shells available at /usr/bin/zsh)
ENV SHELL=/bin/bash \
    DOCKER_BUILDKIT=1

# Mount for docker-in-docker
VOLUME [ "/var/lib/docker" ]

# Fire Docker/Moby script if needed
ENTRYPOINT [ "/usr/local/share/docker-init.sh", "/usr/local/share/ssh-init.sh"]
CMD [ "sleep", "infinity" ]
```

### PR DESCRIPTION
**Pull request by sayhiben**
Fixes #837

### CODE REVIEW

**Comment by sayhiben** [comment]
@microsoft-github-policy-service agree company="Rover.com"

File: src/docker-in-docker/README.md
Comment (samruddhikhandale) [inline src/docker-in-docker/README.md:20]: Once this PR is merged, https://github.com/devcontainers/action will run automatically ; which opens a PR that updates the README file based on `devcontainer-feature.json` and `NOTES.md`.

Hence, can we revert the changes to this file to avoid merge conflicts? Thanks!

File: src/docker-in-docker/install.sh
Comment (samruddhikhandale) [inline src/docker-in-docker/install.sh:13]:
```suggestion
MOBY_BUILDX_VERSION="${MOBYBUILDXVERSION}"
```

Comment (samruddhikhandale) [inline src/docker-in-docker/install.sh:13]: I think it's alright to formalize this Feature option, can you add an entry to https://github.com/devcontainers/features/blob/main/src/docker-in-docker/devcontainer-feature.json ?

**Comment by sayhiben** [comment]
I think I also need to bump the version here, but I'm unsure if you'd consider this a minor or patch bump

File: src/docker-in-docker/install.sh
Comment (samruddhikhandale) [inline src/docker-in-docker/install.sh:13]: Can you bump the minor version for this Feature?

https://github.com/devcontainers/features/blob/f063f4ea83787b34d7c6e97c48b9c84c758a5bcd/src/docker-in-docker/devcontainer-feature.json#L3

Comment (samruddhikhandale) [reply src/docker-in-docker/install.sh:13]: Let's set the `default` value to whichever version currently works. This way folks would have a smooth experience. I'd make sure to remove on Monday or whenever Microsoft fixes it.

Also, add a comment on why set the `default` to a specific version.

File: src/docker-in-docker/install.sh
Comment (samruddhikhandale) [inline src/docker-in-docker/install.sh:210]: How about moving this code blow within `if [ "${USE_MOBY}" = "true" ]; then` ?

**Comment by sayhiben** [comment]
The idea here being that folks can use `"mobyBuildxVersion": "0.12.1-ubuntu20.04u1"` to resolve the current package manifest mismatch through downgrading buildx to a known-good version. This patch will still require that consumers update their devcontainer.json files, so perhaps temporarily pinning would be a better immediate fix?

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]
Thank you for the quick fix, left some feedback

File: src/docker-in-docker/install.sh
Comment (samruddhikhandale) [reply src/docker-in-docker/install.sh:13]: > This patch will still require that consumers update their devcontainer.json files, so perhaps temporarily pinning would be a better immediate fix?

Yep!

**Comment by sayhiben** [comment]
Ah, looks like I need to handle different distributions. Give me a few more mins

**Comment by sayhiben** [comment]
@samruddhikhandale Let's see what GHA has to say about the latest commit. It's a little ugly since it's effectively the same implementation as the suffix-finding code for the docker engine version, but hopefully these can all get a refactor into a common function later ?

**Comment by sayhiben** [comment]
@samruddhikhandale Whoops, had an extra line that'll make the current workflow runs likely fail. Just pushed again

**Comment by samruddhikhandale** [comment]
@sayhiben would be interested in making similar changes once this PR gets approved to 

> @samruddhikhandale Let's see what GHA has to say about the latest commit. It's a little ugly since it's effectively the same implementation as the suffix-finding code for the docker engine version, but hopefully these can all get a refactor into a common function later ?

Yep, this is alright for now, I can refactor it when I remove the `default` value.
Seems like the tests are passing 🎉 

I'll merge this PR when all tests are green. Thank you so much, we appreciate it 👏

**Review by samruddhikhandale — APPROVED** [review]

**Comment by radiantspace** [comment]
Hey, not sure that is same issue or not, but I'm trying to apply the fix and specify moby-buildx version but get errors about package size mismatch when I do so. Relevant logs:

```
#15 14.67 Reading package lists...2024-02-12 14:23:46.153Z: 
#15 15.43 W: Failed to fetch mirror+file:/etc/apt/mirrorlist.txt/dists/focal/InRelease  Could not resolve 'packages.service.github.net'
#15 15.43 W: Some index files failed to download. They have been ignored, or old ones used instead.
#15 15.47 Reading package lists...2024-02-12 14:23:46.894Z: 
2024-02-12 14:23:47.043Z: #15 16.23 Building dependency tree...
#15 16.36 Reading state information...
2024-02-12 14:23:47.162Z: #15 16.48 The following additional packages will be installed:
2024-02-12 14:23:47.293Z: #15 16.48   libdevmapper1.02.1 moby-containerd moby-runc moby-tini
#15 16.48 Recommended packages:
#15 16.48   dmsetup apparmor kmod
#15 16.53 The following NEW packages will be installed:
#15 16.53   libdevmapper1.02.1 moby-buildx moby-cli moby-containerd moby-engine
#15 16.53   moby-runc moby-tini
#15 16.57 0 upgraded, 7 newly installed, 0 to remove and 0 not upgraded.
2024-02-12 14:23:47.294Z: #15 16.57 Need to get 135 MB of archives.
#15 16.57 After this operation, 363 MB of additional disk space will be used.
#15 16.57 Get:1 https://packages.microsoft.com/ubuntu/20.04/prod focal/main amd64 moby-buildx amd64 0.12.1-ubuntu20.04u1 [34.1 MB]
#15 16.57 Err:1 https://packages.microsoft.com/ubuntu/20.04/prod focal/main amd64 moby-buildx amd64 0.12.1-ubuntu20.04u1
#15 16.57   File has unexpected size (34152700 != 34091250). Mirror sync in progress? [IP: 13.80.99.124 443]
#15 16.57   Hashes of expected file:
#15 16.57    - SHA256:0d1c7ac216d2a7d825b6fb0dbdec0f81d88540f1cc074d77abf4c0e7988f3e03
#15 16.57    - Filesize:34091250 [weak]
#15 16.61 Get:2 https://packages.microsoft.com/ubuntu/20.04/prod focal/main amd64 moby-cli amd64 24.0.9-ubuntu20.04u1 [15.6 MB]
2024-02-12 14:23:47.401Z: #15 16.70 Get:3 https://packages.microsoft.com/ubuntu/20.04/prod focal/main amd64 moby-runc amd64 1.1.12-ubuntu20.04u1 [6680 kB]
#15 16.72 Get:4 http://archive.ubuntu.com/ubuntu focal/main amd64 libdevmapper1.02.1 amd64 2:1.02.167-1ubuntu1 [127 kB]
2024-02-12 14:23:47.552Z: #15 16.74 Get:5 https://packages.microsoft.com/ubuntu/20.04/prod focal/main amd64 moby-containerd amd64 1.6.28-ubuntu20.04u1 [45.9 MB]
2024-02-12 14:23:47.680Z: #15 17.00 Get:6 https://packages.microsoft.com/ubuntu/20.04/prod focal/main amd64 moby-tini amd64 0.19.0-ubuntu20.04u1 [350 kB]
2024-02-12 14:23:47.825Z: #15 17.00 Get:7 https://packages.microsoft.com/ubuntu/20.04/prod focal/main amd64 moby-engine amd64 24.0.9-ubuntu20.04u1 [32.2 MB]
#15 17.14 E: Failed to fetch https://packages.microsoft.com/ubuntu/20.04/prod/pool/main/m/moby-buildx/moby-buildx_0.12.1-ubuntu20.04u1_amd64.deb  File has unexpected size (34152700 != 34091250). Mirror sync in progress? [IP: 13.80.99.124 443]
#15 17.14    Hashes of expected file:
#15 17.14     - SHA256:0d1c7ac216d2a7d825b6fb0dbdec0f81d88540f1cc074d77abf4c0e7988f3e03
#15 17.14     - Filesize:34091250 [weak]
#15 17.14 E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
#15 17.14 Fetched 101 MB in 1s (167 MB/s)2024-02-12 14:23:47.826Z: 
2024-02-12 14:23:47.976Z: #15 17.14 (!) Packages for moby not available in OS ubuntu focal (amd64). To resolve, either: (1) set feature option '"moby": false' , or (2) choose a compatible OS version (eg: 'ubuntu-20.04').
#15 17.14 ERROR: Feature "Docker (Docker-in-Docker)" (docker-in-docker) failed to install! Look at the documentation at ********/devcontainers/features/tree/main/src/docker-in-docker for help troubleshooting this error.
2024-02-12 14:23:48.578Z: #15 ERROR: process "/bin/sh -c cp -ar /tmp/build-features-src/docker-in-docker_0 /tmp/dev-container-features  && chmod -R 0755 /tmp/dev-container-features/docker-in-docker_0  && cd /tmp/dev-container-features/docker-in-docker_0  && chmod +x ./devcontainer-features-install.sh  && ./devcontainer-features-install.sh  && rm -rf /tmp/dev-container-features/docker-in-docker_0" did not complete successfully: exit code: 1
2024-02-12 14:23:48.660Z: ------
```

**Pull request merged**
